commcare_connect/opportunity/helpers.py

```diff
@@ -315,25 +314,36 @@ def completed_work_status_total_subquery(status_value):
 
 
 def get_payment_report_data(opportunity: Opportunity):
-    payment_units = PaymentUnit.objects.filter(opportunity=opportunity)
     PaymentReportData = namedtuple(
         "PaymentReportData", ["payment_unit", "approved", "user_payment_accrued", "nm_payment_accrued"]
     )
+
+    report_data_qs = (
+        CompletedWork.objects.filter(opportunity_access__opportunity=opportunity, status=CompletedWorkStatus.approved)
+        .values("payment_unit__name", amount=F("payment_unit__amount"))
+        .annotate(
+            completed_work_count=Count("id"),
+            total_saved_approved=Sum("saved_approved_count"),
+        )
+        .order_by("payment_unit__name")
+    )
     data = []
     total_user_payment_accrued = 0
     total_nm_payment_accrued = 0
-    for payment_unit in payment_units:
-        completed_works = CompletedWork.objects.filter(
-            opportunity_access__opportunity=opportunity, status=CompletedWorkStatus.approved, payment_unit=payment_unit
+    org_pay_per_visit = opportunity.managedopportunity.org_pay_per_visit
+    for group in report_data_qs:
+        user_payment_accrued = group["total_saved_approved"] * group["amount"]
+        nm_payment_accrued = group["completed_work_count"] * org_pay_per_visit
+        data.append(
+            PaymentReportData(
+                group["payment_unit__name"],
+                group["completed_work_count"],
+                user_payment_accrued,
+                nm_payment_accrued,
+            )
         )
-        completed_work_count = len(completed_works)
-        user_payment_accrued = sum([cw.payment_accrued for cw in completed_works])
```

Comment:
This is the key cause of slowness, each call `payment_accrued` fetches all its uservisits to calculate `approved_count` which has been replaced with saved_approved_count above and in DB